examples/vision/metric_learning.py

```diff
@@ -22,290 +24,380 @@
+                      y_display[idx],
+                      nns[idx],
+                      class_mapping=labels,
+                      fig_size=(16, 2))
 
 """
-Finally we run the training. On a Google Colab GPU instance this takes about a minute.
-"""
+# Metrics
 
-model.compile(
-    optimizer=keras.optimizers.Adam(learning_rate=1e-3),
-    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
-)
+We can also plot the extra metrics contained in the `CalibrationResults` to get
+a sense of the matching performance as the distance threshold increases.
 
-history = model.fit(AnchorPositivePairs(num_batchs=1000), epochs=20)
+The following plots show the Precision, Recall, and F1 Score. We can see that
+the matching precision degrades as the distance increases, but that the
+percentage of the queries that we accept as positive matches (recall) grows
+faster up to the calibrated distance threshold.
+"""
 
-plt.plot(history.history["loss"])
+fig, (ax1, ax2) = plt.subplots(1, 2, figsize=(15, 5))
+x = calibration.thresholds['distance']
+
+ax1.plot(x, calibration.thresholds['precision'], label='precision')
+ax1.plot(x, calibration.thresholds['recall'], label='recall')
+ax1.plot(x, calibration.thresholds['f1'], label='f1 score')
+ax1.legend()
+ax1.set_title("Metric evolution as distance increase")
+ax1.set_xlabel('Distance')
+ax1.set_ylim((-0.05, 1.05))
+
+ax2.plot(calibration.thresholds['recall'], calibration.thresholds['precision'])
+ax2.set_title("Precision recall curve")
+ax2.set_xlabel('Recall')
+ax2.set_ylabel('Precision')
+ax2.set_ylim((-0.05, 1.05))
 plt.show()
 
 """
-## Testing
-
-We can review the quality of this model by applying it to the test set and considering
-near neighbours in the embedding space.
-
-First we embed the test set and calculate all near neighbours. Recall that since the
-embeddings are unit length we can calculate cosine similarity via dot products.
+We can also take 100 examples for each class and plot the confusion matrix for
+each example and their nearest match. We also add an "extra" 10th class to
+represent the matches above the calibrated distance threshold.
+
+We can see that most of the errors are between the animal classes with an
+interesting number of confusions between Airplane and Bird. Additionally, we see
+that only a few of the 100 examples for each class returned matches outside of
+the calibrated distance threshold.
 """
 
-near_neighbours_per_example = 10
+cutpoint = 'optimal'
 
-embeddings = model.predict(x_test)
-gram_matrix = np.einsum("ae,be->ab", embeddings, embeddings)
-near_neighbours = np.argsort(gram_matrix.T)[:, -(near_neighbours_per_example + 1) :]
+# This yields 100 examples for each class.
+# We defined this when we created the test_ds sampler.
+x_confusion, y_confusion = test_ds.get_slice(0, -1)
+
+matches = model.match(x_confusion, cutpoint=cutpoint, no_match_label=10)
+confusion_matrix(matches,
+                 y_confusion,
+                 labels=labels,
```

Comment:
Label should be the classes names, I don't think its the case? They are listed below.

Reply:
I updated this as part of https://github.com/tensorflow/similarity/pull/170.

`labels` should now be a list of strs and `class_mapping` should now be the map from class ids to class labels.